### PR TITLE
feat(AJV): Add missing provider level schema properties

### PR DESCRIPTION
**Pull request by fredericbarthelet**
Closes: #8016

### CODE REVIEW

**Comment by fredericbarthelet** [comment]
@medikoo, while working on this PR, I stumbled upon S3 bucket name validation, for which one full service was dedicated (https://github.com/serverless/serverless/blob/master/lib/plugins/aws/lib/validateS3BucketName.js). This service allows use of `.` within bucket's name.

However, as specified by [AWS documentation for bucket name](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html), bucket name including dots won't be able to use S3 transfer acceleration.

Therefore, using a logic similar to the discussion we had regarding schema limitation of lambda env (limited to Ref and GetAtt only since `invoke local` only support those), would you like me to limit bucket name, when specified, to non-doted ones only ?

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -420,14 +509,79 @@ class AwsProvider {
...
+                'ap-southeast-2',
+                'ca-central-1',
+                'eu-central-1',
+                'eu-north-1',
+                'eu-south-1',
+                'eu-west-1',
+                'eu-west-2',
+                'eu-west-3',
+                'me-south-1',
+                'sa-east-1',
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:538]: It seems to miss some regions, e.g. those in China or Canada.

Reference https://docs.aws.amazon.com/general/latest/gr/rande.html

File: lib/plugins/aws/provider/awsProvider.js [hunk 2]
```diff
@@ -405,6 +482,18 @@ class AwsProvider {
...
+                    accessLogging: { type: 'boolean' },
+                    executionLogging: { type: 'boolean' },
+                    format: { type: 'string' },
+                    fullExecutionData: { type: 'boolean' },
+                    level: { enum: ['INFO', 'ERROR'] },
+                    role: { $ref: '#/definitions/awsArn' },
+                    roleManagedExternally: { type: 'boolean' },
+                  },
+                },
                 websocket: {
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:497]: We miss definition for `frameworkLambda`

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @fredericbarthelet ! It looks very promising. Great cleanup, I've picked just few potential issues.

> Therefore, using a logic similar to the discussion we had regarding schema limitation of lambda env (limited to Ref and GetAtt only since invoke local only support those), would you like me to limit bucket name, when specified, to non-doted ones only ?

Good question. I would enable dot, and ensure we throw with meaningful error if it's used and S3 acceleration is enabled (it's opt-in as I see).

Also as I think more about it now, we should have done probably the same to env vars (so allow all formats acceptable for deployment, but throw with meaningful error if local invocation is pursued with not supported env vars - but let's not cover this in this PR)

File: lib/plugins/aws/provider/awsProvider.js [hunk 3]
```diff
@@ -229,6 +229,57 @@ class AwsProvider {
...
+              properties: {
+                Sid: { type: 'string' },
+                Effect: { enum: ['Allow', 'Deny'] },
+                Action: { $ref: '#/definitions/awsIamPolicyAction' },
+                NotAction: { $ref: '#/definitions/awsIamPolicyAction' },
+                Principal: { $ref: '#/definitions/awsIamPolicyPrincipal' },
+                NotPrincipal: { $ref: '#/definitions/awsIamPolicyPrincipal' },
+                Resource: { $ref: '#/definitions/awsIamPolicyResource' },
+                NotResource: { $ref: '#/definitions/awsIamPolicyResource' },
+                Condition: { type: 'object' },
```
[the remark this answers is not in the document]
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js:279]: If you look at the [policy grammar](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_grammar.html), you'll see the following :

image

Sid, NotAction, NotPrincipal, NotResource are listed as part of policy statement. I sourced this schema from https://gist.github.com/jstewmon/ee5d4b7ec0d8d60cbc303cb515272f8a

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/provider/awsProvider.js:279]: Thanks for clarification! You may add this link in comment (after all it's not that straightforward to find precise resource in this case)

File: lib/plugins/aws/provider/awsProvider.js [hunk 4]
```diff
@@ -405,6 +482,18 @@ class AwsProvider {
...
+                  properties: {
+                    accessLogging: { type: 'boolean' },
+                    executionLogging: { type: 'boolean' },
+                    format: { type: 'string' },
+                    fullExecutionData: { type: 'boolean' },
+                    level: { enum: ['INFO', 'ERROR'] },
+                    role: { $ref: '#/definitions/awsArn' },
+                    roleManagedExternally: { type: 'boolean' },
+                  },
+                },
```
[the remark this answers is not in the document]
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js:496]: Understood, however I did not see this part in #8301 , which only partially covers #8018. Would you like me to remove this commit and include it in another PR partially solving the issue, or should I keep it within this PR and update comment to say it partially solve #8018 in addition to close #8016 ?

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/provider/awsProvider.js:496]: In #8301  it was proposed to cover `provider` properties in other PR.

Probably best if you propose it with other PR (also it'll be good to post some note about that at #8301 or #8018)

**Comment by fredericbarthelet** [comment]
> Also as I think more about it now, we should have done probably the same to env vars (so allow all formats acceptable for deployment, but throw with meaningful error if local invocation is pursued with not supported env vars - but let's not cover this in this PR)

S3 acceleration is already disabled when a custom bucketName is provided, wether it includes a dot or not, thanks to the following code block

https://github.com/serverless/serverless/blob/eb5e54847e6e2f6b89a1b5325df4d8421efe479a/lib/plugins/aws/package/lib/generateCoreTemplate.js#L78-L86

The original commit I made allowed for dot within bucket name. Considering your answer, I keep it that way :)

File: lib/plugins/aws/provider/awsProvider.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js]: Understood, removed from this PR

**Comment by medikoo** [comment]
> S3 acceleration is already disabled when a custom bucketName is provided, whether it includes a dot or not, thanks to the following code block

@fredericbarthelet indeed, good point, btw. I've already improved env vars handling with https://github.com/serverless/serverless/pull/8307, so that's covered

File: lib/plugins/aws/provider/awsProvider.js [hunk 5]
```diff
@@ -385,6 +450,27 @@ class AwsProvider {
...
+                    maxPreviousDeploymentArtifacts: { type: 'integer', minimum: 0 },
+                    serverSideEncryption: { enum: ['AES256', 'aws:kms'] },
+                    sseCustomerAlgorithim: { type: 'string' },
+                    sseCustomerKey: { type: 'string' },
+                    sseCustomerKeyMD5: { type: 'string' },
+                    sseKMSKeyId: { type: 'string' },
+                    tags: { $ref: '#/definitions/awsResourceTags' },
+                    blockPublicAccess: { type: 'boolean' },
+                  },
+                },
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:470]: We need `additionalProperties: false`

**Review by medikoo — CHANGES_REQUESTED** [review]
That looks great @fredericbarthelet I've commented just on a few last things, and we should be good to go!

File: lib/plugins/aws/provider/awsProvider.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js:333]: Comment added :)

File: lib/plugins/aws/provider/awsProvider.js [hunk 6]
```diff
@@ -282,6 +282,63 @@ class AwsProvider {
...
+                      { $ref: '#/definitions/awsArn' },
+                      { type: 'array', items: { $ref: '#/definitions/awsArn' } },
+                    ],
+                  },
+                  Federated: {
+                    anyOf: [{ type: 'string' }, { type: 'array', items: { type: 'string' } }],
+                  },
+                  CanonicalUser: {
+                    anyOf: [{ type: 'string' }, { type: 'array', items: { type: 'string' } }],
+                  },
```
[the remark this answers is not in the document]
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js:309]: Nope, I forgot, just added it :)

File: lib/plugins/aws/provider/awsProvider.js [hunk 7]
```diff
@@ -364,9 +429,12 @@ class AwsProvider {
             },
             additionalProperties: false,
           },
-          awsLogGroupName: {
+          awsS3BucketName: {
             type: 'string',
-            pattern: '^[/#A-Za-z0-9-_.]+$',
+            pattern:
+              '(?!^(\\d{1,3}\\.){3}\\d{1,3}$)(^(([a-z0-9]|[a-z0-9][a-z0-9-]*[a-z0-9])\\.)*([a-z0-9]|[a-z0-9][a-z0-9-]*[a-z0-9])$)',
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js:435]: What's the source for this regex?

**Review by medikoo — COMMENTED** [review]
@fredericbarthelet this looks very good, and seems ready to take! I have just one concern which I forgot to add in previous reviews. Let me know.

File: lib/plugins/aws/provider/awsProvider.js (hunk 7 above)
Comment (fredericbarthelet) [reply lib/plugins/aws/provider/awsProvider.js:435]: https://stackoverflow.com/questions/50480924/regex-for-s3-bucket-name

Comment (medikoo) [reply lib/plugins/aws/provider/awsProvider.js:435]: Thanks! Can you add link to this page in comment above this `pattern`. I'm pretty sure others may also wonder about the source of that, and that will help to connect the dots

**Review by medikoo — APPROVED** [review]
Thank you @fredericbarthelet! It's great to have that in

**Pull request merged**